zulipterminal/ui_tools/views.py

```diff
@@ -1593,3 +1594,81 @@ def keypress(self, size: urwid_Size, key: str) -> str:
             )
             return key
         return super().keypress(size, key)
+
+
+class EmojiPickerView(PopUpView):
+    """
+    Displays Emoji Picker view for messages.
+    """
+
+    def __init__(
+        self,
+        controller: Any,
+        title: str,
+        emoji_names: List[str],
+        message: Message,
+        view: Any,
+    ) -> None:
+        self.width = 40
+        self.view = view
+        self.message = message
+        self.controller = controller
+        emoji_buttons = self.generate_emoji_buttons(emoji_names)
+        self.emoji_search = PanelSearchBox(
+            self, "SEARCH_EMOJIS", self.update_emoji_list
+        )
+        search_box = urwid.LineBox(
+            self.emoji_search,
+            tlcorner="─",
+            tline="",
+            lline="",
+            trcorner="─",
+            blcorner="─",
+            rline="",
+            bline="─",
+            brcorner="─",
+        )
+        self.empty_search = False
+        self.allow_update_emoji_list = True
+        self.search_lock = threading.Lock()
+        super().__init__(
+            controller,
+            emoji_buttons,
+            "ADD_REACTION",
+            self.width,
+            title,
+            header=[search_box],
+        )
+
+    @asynch
+    def update_emoji_list(
+        self,
+        search_box: Any = None,
+        new_text: Optional[str] = None,
+        emoji_list: Any = None,
+    ) -> None:
+        pass
+
+    def generate_emoji_buttons(self, emoji_names: List[str]) -> List[EmojiButton]:
+        return [
+            EmojiButton(self.controller, self.width, emoji_name, self.message)
+            for emoji_name in emoji_names
+        ]
+
+    def keypress(self, size: urwid_Size, key: str) -> str:
+        if (
+            is_command_key("ENTER", key)
+            and not self.emoji_search.panel_view.controller.is_in_editor_mode()
+        ):
+            self.allow_update_emoji_list = True
+        if is_command_key("SEARCH_EMOJIS", key) and self.allow_update_emoji_list:
+            self.allow_update_emoji_list = False
+            self.emoji_search.set_edit_text("")
+            self.set_focus("header")
+            return key
+        elif is_command_key("GO_BACK", key):
+            self.emoji_search.reset_search_text()
+            self.allow_update_emoji_list = True
+            self.controller.exit_popup()
+            return key
```

Comment:
Do we need this? `exit_popup` is done by PopupView and we don't need to reset if we are exiting?

Reply:
No, I don't think we need this too. You're right that `exit_popup` is called from PopupView. I added this check earlier when we were not doing "Esc to exit" and didn't bother removing it. I'll remove this in the next iteration. Thanks. :+1:

Reply:
Actually, we do need to keep the `exit_popup` check here, because when `Esc` is pressed from inside emoji search, then the keypress is first handled by `PanelSearchBox` and then by `PopupView`, which throws a ValueError traceback. We need to handle the exit condition here specifically and return so that `PanelSearchBox`'es keypress is not executed after this.